### PR TITLE
fix(telemetry): always create trace id even if the trace is not sampled

### PR DESCRIPTION
**Pull request by bnjjj**
Now when using `experimental_response_trace_id` for example, you'll always have a valid trace_id in response headers even if the trace is not sampled. It's following the behavior of the [OpenTelemetry specs](https://opentelemetry.io/docs/specs/otel/trace/sdk/#sampling). 

close https://github.com/apollographql/router/issues/3208

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/dynamic_attribute.rs [hunk 1]
```diff
@@ -223,22 +223,26 @@ impl EventDynAttribute for ::tracing::Span {
...
+                                    // Can't use ::tracing::error! because it could create deadlock on extensions
+                                    eprintln!("no OtelData, this is a bug");
                                 }
-                            },
-                            None => {
-                                // Can't use ::tracing::error! because it could create deadlock on extensions
-                                eprintln!("no OtelData, this is a bug");
                             }
+                        } else {
+                            // FIXME: we should put event attributes somewhere else to make it work even if it's not sampled like we did with LogAttributes
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/dynamic_attribute.rs:245]: I created the [related issue ](https://github.com/apollographql/router/issues/5448)

File: apollo-router/src/plugins/telemetry/dynamic_attribute.rs [hunk 2]
```diff
@@ -258,23 +262,27 @@ impl EventDynAttribute for ::tracing::Span {
...
```
Comment (bnjjj) [inline apollo-router/src/plugins/telemetry/dynamic_attribute.rs:285]: Related [issue](https://github.com/apollographql/router/issues/5448) created

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Can we add some tests to verify that the traceID is making it to the logging also?
Either in https://github.com/apollographql/router/blob/dev/apollo-router/tests/integration/telemetry/logging.rs or the jaeger tests.

**Comment by bnjjj** [comment]
@BrynCooke Added 2 tests for json and text format with sampler set to `always_off` to make sure it works when it's sampled and not sampled https://github.com/apollographql/router/pull/5447/commits/e81d7cdfe7e0788499b2537fd78f4fe9bfedda25

**Comment by bnjjj** [comment]
@BrynCooke Correct me if I'm wrong but I think this one will fix https://github.com/apollographql/router/issues/5387 right ?

**Review by garypen — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/otel/layer.rs [hunk 3]
```diff
@@ -650,24 +757,53 @@ where
...
+                Some(SampledSpan::Sampled(trace_id, _) | SampledSpan::NotSampled(trace_id, _)) => {
+                    opentelemetry_api::trace::TraceId::from(trace_id.to_u128())
+                }
+                // It's probably the root span
+                None => self.tracer.new_trace_id(),
+            }
+        };
+
+        let span_id = self.tracer.new_span_id();
+        let sampled = if is_sampled {
```
Comment (BrynCooke) [inline apollo-router/src/plugins/telemetry/otel/layer.rs:784]: Nice!

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/config_new/events.rs [hunk 4]
```diff
@@ -278,23 +289,36 @@ impl Instrumented
...
+                ),
+            ));
+            attrs.push(KeyValue::new(
+                Key::from_static_str("http.request.body"),
+                opentelemetry::Value::String(
+                    serde_json::to_string(request.supergraph_request.body())
+                        .unwrap_or_default()
+                        .into(),
+                ),
+            ));
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/config_new/events.rs:321]: Events will always be displayed even if the span is not sampled. But I think one of improvement we can do here is to check if the user set the level `debug` we can know thanks to a helper in `tracing` I think and directly not collecting anything related to that event.

**Pull request merged**